Draft "common issues" doc; update production guidance

taroface opened the pull request:
Fixes DOC-90
Fixes DOC-383
Fixes DOC-373
Fixes DOC-690
Fixes DOC-793
Contributes to DOC-498; not sure if the fix will be satisfactory

- Draft a new document, "Common Issues to Monitor", collecting guidance from the [Most Common Problems runbook](https://github.com/cockroachlabs/cockroachdb-runbook-template/tree/main/most-common-problems)
- Update Troubleshooting and Production Checklist docs accordingly
- Move many common pieces of guidance into includes
- Clarify and add example on memory management and disk spilling

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/docs/12687)
<!-- Reviewable:end -->

taroface commented on _includes/v21.2/prod-deployment/cluster-unavailable-monitoring.md:
```diff
@@ -0,0 +1,3 @@
+{{site.data.alerts.callout_info}}
+If the cluster becomes unavailable, the DB Console and Cluster API will also become unavailable. You can continue to monitor the cluster via the [Prometheus endpoint](monitoring-and-alerting.html#prometheus-endpoint) and [logs](logging-overview.html).
```
Will mention pprof endpoints once these are documented with DOC-2183

taroface commented on _includes/v21.2/prod-deployment/healthy-lsm.md:
```diff
@@ -0,0 +1 @@
+**Expected values for a healthy cluster**: The number of L0 files should be fewer than 1000. High values indicate heavy write load that is causing accumulation of files in level 0. These files are not being compacted quickly enough to lower levels, resulting in a misshapen LSM.
```
"Fewer than 1000" is based on the current threshold at which admission control takes over. Not sure if we have better guidance on this yet.

taroface replied:
Ballpark is correct. Emphasize order of magnitude. Files in the hundreds to low thousands is OK for bigger servers.

High thousands are not good.

There is additional nuance here relating to vCPU.

taroface commented on _includes/v21.2/prod-deployment/healthy-read-amplification.md:
```diff
@@ -0,0 +1 @@
+**Expected values for a healthy cluster**: Read amplification factor should be in the single digits. A value higher than 50 strongly suggests that the LSM tree has an unhealthy shape.
```
"Higher than 50" was gleaned from this [postmortem](https://cockroachlabs.atlassian.net/wiki/spaces/TS/pages/1548484769/2021-03-06+Postmortem+on+DevSisters+7788)

taroface replied:
See: https://github.com/cockroachlabs/cockroachdb-runbook-template/blob/main/monitoring-alerts/alert-lsm-health.md

Add nuance on duration of high read amp value

taroface commented on _includes/v21.2/prod-deployment/healthy-sql-memory.md:
```diff
@@ -0,0 +1 @@
+**Expected values for a healthy cluster**: This value should not exceed the [`--max-sql-memory`](recommended-production-settings.html#cache-and-sql-memory-size) size.
```
I'm not sure this is correct or detailed enough - my understanding is that `--max-sql-memory` sets a soft limit and can be exceeded. But will that always result in a `memory budget exceeded` error, or does it happen more routinely?

taroface commented on _includes/v21.2/prod-deployment/healthy-storage-capacity.md:
```diff
@@ -0,0 +1 @@
+**Expected values for a healthy cluster**: Used capacity should be no more than 80% of the total capacity.
```
I am pretty sure this 80% is an arbitrary value I picked up through previous threads - unfortunately can't find the exact source now. Do we have more precise guidance? Also, this differs from our Prometheus alert that is preconfigured to alert at less than 15% available capacity.

taroface replied:
10-15% would be critical. 20% a warning. Also see: https://github.com/cockroachlabs/cockroachdb-runbook-template/blob/main/monitoring-alerts/alert-node-storage-capacity.md

taroface commented on v21.2/cluster-setup-troubleshooting.md:
```diff
@@ -284,30 +284,38 @@ See the following FAQs:
 RAM capacity | Memory Usage | Consistently more than 80%
-CPU capacity | CPU Percent | Consistently less than 20% in idle (i.e.:80% busy)
+CPU capacity | CPU Percent | Consistently less than 20% in idle (i.e., 80% busy)
+Disk capacity | Available Disk Capacity | Consistently less than 20% of the [store](cockroach-start.html#store) size
```
As I mentioned above: I am pretty sure this is an arbitrary value I picked up through previous threads - unfortunately can't find the exact source now. Do we have more precise guidance?

taroface commented on v21.2/common-errors.md:
```diff
@@ -187,6 +188,22 @@ the user.
+
+For [disk-spilling operations](vectorized-execution.html#disk-spilling-operations) such as hash joins that are memory-intensive, another solution is to increase the `sql.distsql.temp_storage.workmem` [cluster setting](cluster-settings.html) to allocate more memory to the operation before it spills to disk and likely consumes more memory. This improves the performance of the query, though at a pos ...
+
+For example, if a query contains a hash join that requires 128 MiB of memory before spilling to disk, values of `sql.distsql.temp_storage.workmem=64MiB` and `--max-sql-memory=1GiB` allow the query to run with a concurrency of 16 without errors. The 17th concurrent instance will exceed `--max-sql-memory` and produce a `memory budget exceeded` error. Increasing `sql.distsql.temp_storage.workmem` to ...
```
@yuzefovich, not sure if I correctly interpreted the `sql.distsql.temp_storage.workmem` part of your example in the bottom paragraph here.

taroface commented on v21.2/common-issues-to-monitor.md:
```diff
@@ -0,0 +1,292 @@
+| [Node heartbeat latency](#node-heartbeat-latency) | The time between [node liveness](cluster-setup-troubleshooting.html#node-liveness-issues) heartbeats.                |
+| [Command commit latency](#command-commit-latency) | The speed at which [Raft commands](architecture/replication-layer.html) are being committed by nodes in the cluster. |
+
+#### Storage capacity
```
@a-entin, although it wasn't part of your Most Common Problems runbook, I also added this storage capacity section so that all the sizing guidance can be together. Let me know if we should make additional points here.

taroface commented on v21.2/cluster-setup-troubleshooting.md:
```diff
@@ -490,56 +513,18 @@ Common reasons for node liveness issues include:
-
-**Expected values for a healthy cluster**: Less than 100ms in addition to the network latency between nodes in the cluster.
-
-### Check node liveness record last update
```
I removed this section (just the "node liveness record last update" part) because, following a convo with @dhartunian, this seems to be checking the last update to node **status** in the DB Console, not liveness record. I can make a correction in case we do have a way to check for the latter.

taroface commented:
@a-entin For this update, I focused on the CPU, memory, and storage/disk I/O guidance in your runbook. The highlights are:

- There is a new doc: https://deploy-preview-12687--cockroachdb-docs.netlify.app/docs/stable/common-issues-to-monitor.html
- Which led to some updates to the troubleshooting doc, starting at: https://deploy-preview-12687--cockroachdb-docs.netlify.app/docs/stable/cluster-setup-troubleshooting.html#capacity-planning-issues

@yuzefovich I folded DOC-383 into this PR; sorry there are a lot of unrelated updates here. The bulk of the pertinent changes are at:

- https://deploy-preview-12687--cockroachdb-docs.netlify.app/docs/stable/common-errors.html#memory-budget-exceeded
- https://deploy-preview-12687--cockroachdb-docs.netlify.app/docs/v21.2/vectorized-execution#disk-spilling-operations

yuzefovich reviewed (commented):
> Reviewed 1 of 54 files at r1, 1 of 6 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/docs/12687)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @a-entin, @taroface, and @thtruo)

___
*[_includes/v21.2/prod-deployment/healthy-sql-memory.md, line 1 at r2](https://reviewable.io/reviews/cockroachdb/docs/12687#-MspRFp4ItPRHmNhdsVs-r2-1:-Mt4kU7q7ea2m-XZtcwu:b-9hzzrx) ([raw file](https://github.com/cockroachdb/docs/blob/5b9458e2dc907f0b2a527020a8ffbcf3c4abb695/_includes/v21.2/prod-deployment/healthy-sql-memory.md#L1)):*
<details><summary><i>Previously, taroface (Ryan Kuo) wrote…</i></summary><blockquote>

I'm not sure this is correct or detailed enough - my understanding is that `--max-sql-memory` sets a soft limit and can be exceeded. But will that always result in a `memory budget exceeded` error, or does it happen more routinely? cc @yuzefovich 
</blockquote></details>

Sorry, I'm having trouble understanding what `This value` refers to, could you please link me to that spot?

`--max-sql-memory` sets a _hard_ limit on the memory usage that is accounted for. Whenever a CRDB node performs some memory allocation, it is not accounted for with our memory monitoring system, so some allocations are not tracked (in some cases this is undesirable, in others it is a problem that we're slowly addressing). However, all allocations that are tracked - when combined - will never exceed `--max-sql-memory`, and whenever the tracking would put the accounted memory usage above the limit, `memory budget exceeded` error is returned instead.
___
*[v21.2/common-errors.md, line 201 at r2](https://reviewable.io/reviews/cockroachdb/docs/12687#-MspRFp4ItPRHmNhdsWG-r2-201:-Mt4lbaN7JU-01cOx5rC:b-xl8xz2) ([raw file](https://github.com/cockroachdb/docs/blob/5b9458e2dc907f0b2a527020a8ffbcf3c4abb695/v21.2/common-errors.md#L201)):*
<details><summary><i>Previously, taroface (Ryan Kuo) wrote…</i></summary><blockquote>

@yuzefovich, not sure if I correctly interpreted the `sql.distsql.temp_storage.workmem` part of your example in the bottom paragraph here.
</blockquote></details>

Yes, this sounds right.


<!-- Sent from Reviewable.io -->

taroface commented:
> Sorry, I'm having trouble understanding what `This value` refers to, could you please link me to that spot?

Yep, it refers to the SQL Memory time-series graph: https://deploy-preview-12687--cockroachdb-docs.netlify.app/docs/v21.2/ui-sql-dashboard#sql-memory

It seems this will never exceed `--max-sql-memory` by definition? If so, is there guidance we can provide on what a healthy threshold would be?

yuzefovich reviewed (commented):
I see, thanks. Indeed, it should never exceed `--max-sql-memory` by definition. In terms of healthy threshold, `75%` or `80%` of `--max-sql-memory` seem good to me.

Changes to
- https://deploy-preview-12687--cockroachdb-docs.netlify.app/docs/stable/common-errors.html#memory-budget-exceeded
- https://deploy-preview-12687--cockroachdb-docs.netlify.app/docs/v21.2/vectorized-execution#disk-spilling-operations

LGTM.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/docs/12687#-:-Mt4wHs89cdvoO8XNn5D:btsg9wu)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @a-entin, @taroface, and @thtruo)



<!-- Sent from Reviewable.io -->

taroface commented:
@a-entin I've updated following our chat the other day. Would you mind having another look at the following:

- The "Common Issues" page: https://deploy-preview-12687--cockroachdb-docs.netlify.app/docs/stable/common-issues-to-monitor.html
- New "Sizing" section in Production Checklist (this removes/replaces the "32 vCPU sweet spot" messaging): https://deploy-preview-12687--cockroachdb-docs.netlify.app/docs/v21.2/recommended-production-settings#sizing
- "Capacity Planning Issues", slightly updated from existing docs: https://deploy-preview-12687--cockroachdb-docs.netlify.app/docs/v21.2/cluster-setup-troubleshooting#capacity-planning-issues
- Troubleshooting insufficient CPU: https://deploy-preview-12687--cockroachdb-docs.netlify.app/docs/v21.2/cluster-setup-troubleshooting#cpu-is-insufficient-for-the-workload
- Troubleshooting OOM crash: https://deploy-preview-12687--cockroachdb-docs.netlify.app/docs/v21.2/cluster-setup-troubleshooting#out-of-memory-oom-crash

a-entin commented:
@taroface, a few comments re content in `deploy-preview-12687--cockroachdb-docs`:
>For good performance, we recommend at least 8 vCPUs per node
As noted earlier in that page, using larger nodes brings stability benefits, rather than performance gains. For the same total vcpus in a cluster, using fewer larger nodes makes the cluster more resilient to hot spots and therefore more robust.

>A smaller node will also take a shorter time to rebuild
I presume "rebuild" refers to a smaller node is quicker to rebalance to a steady state. true. perhaps worth clarifying.

> "cores" refer to vCPUs, which are also known as hyperthreads.
I suggest we just say that here (this doc page) a vCPU maps to a whole core in the underlying hardware platform.
It's better to not mention hyperthreading at all, sort of implying hyperthreading should be at left at its default setting in the bios. The reason for this suggestion is we have no studies and no guidance if hyperthreading is beneficial for crdb or not. Note that a vcpu can be a fraction of a core, but our sizing and practices are based on 1:1 vcpu:core.

> Outright I/O failure due to a disk stall
An "outright failure" is a hardware malfunction. A disk "stall" is a temporary performance issue. Mutually exclusive, one does not imply the other. Consider deleting this line. The main reason is our "disk stall" message in the log is very misleading. In a majority of situations it's a result of a cpu overload rather than a disk overload.

> Guidance in `https://deploy-preview-12687--cockroachdb-docs.netlify.app/docs/v21.2/recommended-production-settings#aws` is mostly incorrect and may need to be revised entirely.
For example: The recommended VM type is `m5`. It has 4GB/vcpu we recommend. `c5` only has 2GB/vcpu and it is not suitable for prod. Ccloud standardized on m5.
`gp2` storage is not suitable for crdb, period. We should be consistent with CCloud where the AWS EBS storage type is `gp3`.

taroface commented:
> > For good performance, we recommend at least 8 vCPUs per node
> 
> As noted earlier in that page, using larger nodes brings stability benefits, rather than performance gains. For the same total vcpus in a cluster, using fewer larger nodes makes the cluster more resilient to hot spots and therefore more robust.

Replaced "performance" with "stability." I think the (my) confusion was with the original wording of "performance consistency" to describe "cluster stability". I've updated this as well.
 
> > A smaller node will also take a shorter time to rebuild
> 
> I presume "rebuild" refers to a smaller node is quicker to rebalance to a steady state. true. perhaps worth clarifying.

Fixed, using this wording.

> > "cores" refer to vCPUs, which are also known as hyperthreads.
> 
> I suggest we just say that here (this doc page) a vCPU maps to a whole core in the underlying hardware platform. It's better to not mention hyperthreading at all, sort of implying hyperthreading should be at left at its default setting in the bios. The reason for this suggestion is we have no studies and no guidance if hyperthreading is beneficial for crdb or not. Note that a vcpu can be a fraction of a core, but our sizing and practices are based on 1:1 vcpu:core.

Replaced with "In our sizing and production guidance, 1 vCPU is considered equivalent to 1 core in the underlying hardware platform."

> > Outright I/O failure due to a disk stall
> 
> An "outright failure" is a hardware malfunction. A disk "stall" is a temporary performance issue. Mutually exclusive, one does not imply the other. Consider deleting this line. The main reason is our "disk stall" message in the log is very misleading. In a majority of situations it's a result of a cpu overload rather than a disk overload.

I removed "Outright I/O failure due to". Since this references an entire section on "Disk stalls" that long predates me or this PR, I think it's best to file a separate issue to revise that section, if it's incorrect or misleading.

> > Guidance in `https://deploy-preview-12687--cockroachdb-docs.netlify.app/docs/v21.2/recommended-production-settings#aws` is mostly incorrect and may need to be revised entirely.
> 
> For example: The recommended VM type is `m5`. It has 4GB/vcpu we recommend. `c5` only has 2GB/vcpu and it is not suitable for prod. Ccloud standardized on m5. `gp2` storage is not suitable for crdb, period. We should be consistent with CCloud where the AWS EBS storage type is `gp3`.

This update is being handled in an [open PR](https://github.com/cockroachdb/docs/pull/12445). I will comment with the above and see if we can merge it soon.

a-entin reviewed (approved):
LGTM

kathancox commented on cockroachcloud/alerts-page.md:
```diff
@@ -7,7 +7,7 @@ toc: true
 
 {{site.data.alerts.callout_info}}
-The Alerts page is accessible on {{ site.data.products.dedicated }} clusters. For {{ site.data.products.serverless }} cluters, all [Console Admins](console-access-management.html#console-admin) will automatically receive email alerts when your cluster reaches 75% and 100% of its burst capacity or storage limit. If you set a spend limit, you will also receive alerts at 50%, 75%, and 100% of your s ...
+The Alerts page is accessible on {{ site.data.products.dedicated }} clusters. For {{ site.data.products.serverless }} clusters, all [Console Admins](console-access-management.html#console-admin) will automatically receive email alerts when your cluster reaches 75% and 100% of its burst capacity or storage limit. If you set a spend limit, you will also receive alerts at 50%, 75%, and 100% of your  ...
```
Looks like this was just a typo correction, but the "Alerts page" in this note could be bold like the previous paragraph, if it's UI?

kathancox commented on v21.2/cluster-setup-troubleshooting.md:
```diff
@@ -362,7 +370,7 @@ Symptoms of disk stalls include:
 Causes of disk stalls include:
 
-- Disk operations have backed up due to underprovisioned IOPS. Make sure you are deploying with our [recommended production settings for storage](recommended-production-settings.html#storage).
+- Disk operations have backed up due to underprovisioned IOPS. Make sure you are deploying with our [recommended production settings for storage](recommended-production-settings.html#storage) and [monitoring disk IOPS](common-issues-to-monitor.html#disk-iops).
```
This is a super nit (and it might be because I always have backups on the brain), but here "backed up" is meaning slowed or in some kind of queue? I wondered if there was possibility for confusion there — in terminology. But again, might just be me!

kathancox commented on v21.2/cluster-setup-troubleshooting.md:
```diff
@@ -376,59 +384,79 @@ CockroachDB's built-in disk stall detection works as follows:
-- **CGo allocated memory**: Memory allocated by the C/C++ libraries linked into CockroachDB and primarily concerns the block caches for the storage engine (this is true for both [Pebble (the default engine) and RocksDB](cockroach-start.html#storage-engine)). This is the “cache” mentioned in the note above. The size of CGo allocated memory is usually very close to the configured cache size.
-- **Overhead**: The process resident set size minus Go/CGo allocated memory.
+- **Go allocated memory**: Memory allocated by the Go runtime to support query processing and various caches maintained in Go by CockroachDB.
+- **CGo allocated memory**: Memory allocated by the C/C++ libraries linked into CockroachDB and primarily concerns the block caches for the [Pebble storage engine](cockroach-start.html#storage-engine)). This is the "cache" mentioned in the note above. The size of CGo allocated memory is usually very close to the configured cache size.
```
Maybe avoid "above" as a direction? I wonder if this would be better as an anchor link? Because I'm not sure which note is meant here.

kathancox commented on v21.2/recommended-production-settings.md:
```diff
@@ -89,21 +113,23 @@ We recommend provisioning volumes with **150 GiB per vCPU**. It's fine to have l
 
 {{site.data.alerts.callout_info}}
-Underprovisioning storage leads to node crashes when the disks fill up. Once this has happened, it is difficult to recover from. To prevent your disks from filling up, provision enough storage for your workload, monitor your disk usage, and use a ballast file as described above. For more information, see [capacity planning issues](cluster-setup-troubleshooting.html#capacity-planning-issues) and [ ...
+Under-provisioning storage leads to node crashes when the disks fill up. Once this has happened, it is difficult to recover from. To prevent your disks from filling up, provision enough storage for your workload, monitor your disk usage, and use a ballast file as described above. For more information, see [capacity planning issues](cluster-setup-troubleshooting.html#capacity-planning-issues) and  ...
```
With "above" here might want to be more specific about where this is described? I assume in the two preceding paragraphs.

kathancox commented on v21.2/recommended-production-settings.md:
```diff
@@ -21,66 +21,90 @@ We recommend running a [glibc](https://www.gnu.org/software/libc/)-based Linux d
-- Avoid "burstable" or "shared-core" virtual machines that limit the load on CPU resources.
+    {{site.data.alerts.callout_info}} 
+    The benefits to having more RAM decrease as the [number of vCPUs](#sizing) increases.
+    {{site.data.alerts.end}}
```
This note is indented the next note is not indented with the bullet point, is that purposeful?

kathancox reviewed (approved):
LGTM, a few little nits!

taroface reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/docs/12687)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @kathancox, @taroface, and @thtruo)

___
*[_includes/v21.2/prod-deployment/healthy-lsm.md, line 1 at r8](https://reviewable.io/reviews/cockroachdb/docs/12687#-MspRFp4ItPRHmNhdsVo-r8-1:-MuR79VvER5_A9ZBi5ld:b-fl6crf) ([raw file](https://github.com/cockroachdb/docs/blob/709a33a184a178b40eb196531f040143f5195898/_includes/v21.2/prod-deployment/healthy-lsm.md#L1)):*
<details><summary><i>Previously, kathancox (Kathryn Hancox) wrote…</i></summary><blockquote>

nit: the style guide favors bold over italics. I see there is bold at the start of the sentence... I think the **not** will still show?
</blockquote></details>

This is news to me! Fixed and thanks for pointing this out.
___
*[cockroachcloud/alerts-page.md, line 10 at r8](https://reviewable.io/reviews/cockroachdb/docs/12687#-MspRFp4ItPRHmNhdsWD-r8-10:-MuR7qGNEIsOjgJYr7kY:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/709a33a184a178b40eb196531f040143f5195898/cockroachcloud/alerts-page.md#L10)):*
<details><summary><i>Previously, kathancox (Kathryn Hancox) wrote…</i></summary><blockquote>

Looks like this was just a typo correction, but the "Alerts page" in this note could be bold like the previous paragraph, if it's UI?
</blockquote></details>

Done.
___
*[v21.2/cluster-setup-troubleshooting.md, line 373 at r8](https://reviewable.io/reviews/cockroachdb/docs/12687#-MspRFp4ItPRHmNhdsWE-r8-373:-MuR80Ef5F_zf5Hr4zcV:btn1bii) ([raw file](https://github.com/cockroachdb/docs/blob/709a33a184a178b40eb196531f040143f5195898/v21.2/cluster-setup-troubleshooting.md#L373)):*
<details><summary><i>Previously, kathancox (Kathryn Hancox) wrote…</i></summary><blockquote>

This is a super nit (and it might be because I always have backups on the brain), but here "backed up" is meaning slowed or in some kind of queue? I wondered if there was possibility for confusion there — in terminology. But again, might just be me! 
</blockquote></details>

I agree. Changed to "slowed"!
___
*[v21.2/cluster-setup-troubleshooting.md, line 418 at r8](https://reviewable.io/reviews/cockroachdb/docs/12687#-MspRFp4ItPRHmNhdsWE-r8-418:-MuR8Ejy3i7jXLM0kJdu:b7wfozf) ([raw file](https://github.com/cockroachdb/docs/blob/709a33a184a178b40eb196531f040143f5195898/v21.2/cluster-setup-troubleshooting.md#L418)):*
<details><summary><i>Previously, kathancox (Kathryn Hancox) wrote…</i></summary><blockquote>

Maybe avoid "above" as a direction? I wonder if this would be better as an anchor link? Because I'm not sure which note is meant here. 
</blockquote></details>

I clarified this wording. At some point these docs should be revamped. This is a fairly old section that dates back to a wiki page I found.
___
*[v21.2/recommended-production-settings.md, line 79 at r8](https://reviewable.io/reviews/cockroachdb/docs/12687#-MspRFp4ItPRHmNhdsWQ-r8-79:-MuR93w4CqhYQiI6x0FR:b2n22ia) ([raw file](https://github.com/cockroachdb/docs/blob/709a33a184a178b40eb196531f040143f5195898/v21.2/recommended-production-settings.md#L79)):*
<details><summary><i>Previously, kathancox (Kathryn Hancox) wrote…</i></summary><blockquote>

This note is indented the next note is not indented with the bullet point, is that purposeful? 
</blockquote></details>

It was, but I see the intent was lost after some things were moved around. Fixed.
___
*[v21.2/recommended-production-settings.md, line 116 at r8](https://reviewable.io/reviews/cockroachdb/docs/12687#-MspRFp4ItPRHmNhdsWQ-r8-116:-MuR9RWeDdjV1m8upHmv:b7jdgth) ([raw file](https://github.com/cockroachdb/docs/blob/709a33a184a178b40eb196531f040143f5195898/v21.2/recommended-production-settings.md#L116)):*
<details><summary><i>Previously, kathancox (Kathryn Hancox) wrote…</i></summary><blockquote>

With "above" here might want to be more specific about where this is described? I assume in the two preceding paragraphs. 
</blockquote></details>

Using a link instead!


<!-- Sent from Reviewable.io -->
